Core/include/Acts/Seeding/CandidatesForSpM.hpp

```diff
@@ -9,141 +9,136 @@
-  const typename CandidatesForSpM<external_space_point_t>::sp_type&
-  CandidatesForSpM<external_space_point_t>::spM() const
-  { return m_SpM; }
-  
-  template<typename external_space_point_t>
-  inline void CandidatesForSpM<external_space_point_t>::setMaxElements(std::size_t n_low,
-								       std::size_t n_high)
-  {
-    if (m_storage_high.capacity() < n_high) m_storage_high.reserve(n_high);
-    if (m_storage_low.capacity() < n_low) m_storage_low.reserve(n_low);
-    m_max_size_high = n_high;
-    m_max_size_low = n_low;
-  }
-
-  template<typename external_space_point_t>
-  inline void CandidatesForSpM<external_space_point_t>::setMediumSp(typename CandidatesForSpM<external_space_point_t>::sp_type idx)
-  { m_SpM = idx; }
-
-  template<typename external_space_point_t>
-  inline void CandidatesForSpM<external_space_point_t>::setBottomSp(typename CandidatesForSpM<external_space_point_t>::sp_type idx)
-  { m_SpB = idx; }
-
-  template<typename external_space_point_t>
-  inline float CandidatesForSpM<external_space_point_t>::top(const std::vector<value_type>& storage) const
-  { return weight(storage, 0); }
-
-  template<typename external_space_point_t>
-  inline bool CandidatesForSpM<external_space_point_t>::exists(std::size_t n, std::size_t max_size) const
-  { return n < max_size; }
-
-  template<typename external_space_point_t>
-  inline float CandidatesForSpM<external_space_point_t>::weight(const std::vector<value_type>& storage, std::size_t n) const
-  { return std::get<InternalComponents::I_WEIGHT>(storage[n]); }
-
-  template<typename external_space_point_t>
-  inline void CandidatesForSpM<external_space_point_t>::clear()
-  {
-    // do not clear max size, this is set only once
-    m_n_high = 0;
-    m_n_low = 0;
-    // clean fixed space points
-    m_SpB = default_value;
-    m_SpM = default_value;
-    // clean storage
-    m_storage_high.clear();
-    m_storage_low.clear();
-  }
-  
+template <typename external_space_point_t>
+class CandidatesForSpM {
+  // Internal collection, we store only the components that are different and
+  // useful
+  enum InternalComponents : int { I_BSP = 0, I_TSP, I_WEIGHT, I_ZORIGIN };
+
+ public:
+  // more complete collection of variables, used by external seeding code
+  enum Components : int { BSP = 0, MSP, TSP, WEIGHT, ZORIGIN, QUALITY };
+
+  using sp_type = external_space_point_t*;
+  using value_type = std::tuple<sp_type, sp_type, float, float>;
+  using output_type = std::tuple<sp_type, sp_type, sp_type, float, float, bool>;
+  static constexpr sp_type default_value = nullptr;
+
+  CandidatesForSpM();
+  ~CandidatesForSpM() = default;
+
+  void setMaxElements(std::size_t n_low, std::size_t n_high);
+  void setMediumSp(sp_type);
+  void setBottomSp(sp_type);
+  std::vector<output_type> storage() const;
+  const sp_type& spM() const;
+
+  void push(sp_type& SpT, float weight, float zOrigin, bool isQuality);
+  void clear();
+
+ private:
+  bool exists(std::size_t, std::size_t) const;
```

Comment:
unclear what the meaning of the arguments is

Reply:
It verifies the storage has that element available. That is used to determine if a father node has a valuable child stored in the collection. The first argument is the node number we are asking to retrieve, the second argument is the number of "valid" elements (i.e. `m_n`).

In the current implementation the `storage` has a maximum size of `m_max_size`, while the number of collected elements in the `storage` is given by `m_n`. When we have added `m_max_size` elements in the collection (and `storage.size()` returns `m_max_size`) and we pop one of them we simply reduce `m_n` by one to save time: the popped element is still in the collection but is considered as "non valid". That means that `storage.size()` will keep returning `m_max_size` but only a few of them are "valid" (i.e. all those with an index lower then `m_n`).

The `exists` function checks we are not trying to access a non valid element.